packages/nextjs/services/database/bg-imports/0002_batch_data.ts

```diff
@@ -24,74 +24,70 @@ async function importData() {
-          console.log(`Updated existing batch: ${createdBatch.name} (ID: ${createdBatch.id})`);
-        } else {
-          const result = await tx.insert(batches).values(newBatch).returning();
-          createdBatch = result[0];
-          console.log(`Imported batch: ${createdBatch.name} (ID: ${createdBatch.id})`);
-        }
-
-        // Store the mapping between Firebase ID and our database ID
-        batchIdMap.set(batch.name, createdBatch.id);
+    // Save the mapping between Firebase ID and our database ID
+    const batchIdMap = new Map<string, number>();
+
+    // Preload existing batches
+    const existingBatches = await db.select().from(batches);
+    const existingBatchesByTelegramLink: Record<string, (typeof existingBatches)[0]> = {};
+    existingBatches.forEach(batch => {
+      if (batch.telegramLink) {
+        existingBatchesByTelegramLink[batch.telegramLink] = batch;
       }
+    });
 
-      // Update users with batch information
-      for (const user of usersData) {
-        const batchId = batchIdMap.get(user.batch.number);
-
-        if (!batchId) {
-          console.warn(`Batch ${user.batch.number} not found for user ${user.id}`);
-          continue;
-        }
-
-        // Use transaction for update
-        const result = await tx
-          .update(users)
-          .set({
-            batchId: batchId,
-            batchStatus: user.batch.status.toLowerCase() as BatchUserStatus,
-          })
-          .where(eq(users.userAddress, user.id));
-
-        if (result.rowCount === 0) {
-          console.warn(`User ${user.id} not found in database`);
-          continue;
-        }
-
-        console.log(`Updated user: ${user.id} with batch ${user.batch.number}`);
+    // Import batches
+    for (const batch of batchesData) {
+      const newBatch = {
+        name: `Batch ${batch.name}`,
+        startDate: new Date(batch.startDate),
+        status: batch.status.toLowerCase() as BatchStatus,
+        contractAddress: batch.contractAddress,
+        telegramLink: batch.telegramLink,
+        bgSubdomain: `batch${batch.name}`,
+      };
+
+      let createdBatch;
+      if (batch.telegramLink && batch.telegramLink in existingBatchesByTelegramLink) {
+        // Update existing batch with any new information
+        const existingBatch = existingBatchesByTelegramLink[batch.telegramLink];
+        const result = await db.update(batches).set(newBatch).where(eq(batches.id, existingBatch.id)).returning();
+        createdBatch = result[0];
+        console.log(`Updated existing batch: ${createdBatch.name} (ID: ${createdBatch.id})`);
+      } else {
+        const result = await db.insert(batches).values(newBatch).returning();
+        createdBatch = result[0];
+        console.log(`Imported batch: ${createdBatch.name} (ID: ${createdBatch.id})`);
       }
-    });
+
+      // Store the mapping between Firebase ID and our database ID
+      batchIdMap.set(batch.name, createdBatch.id);
+    }
```

Comment:
Soo here the reason why we can't use batch is because, we wanted the result of the inserted batch (batch id) 


We cab kind of achieve this with `batch` since batch also returns data but its very hard to get the types right. It's very dynamic and also the way we are constructing the queries are very dynamic too. 

So I kind of falling to using the transaction for batches script and also kind of makes sense to use but let me know 🙌

Reply:
Thanks for exploring this, Shiv!

> So I kind of falling to using the transaction for batches script and also kind of makes sense to use but let me know 🙌

I agree with this!

I think transactions make sense here (+ getting the rolling back out of the box). Also, this is little data, so no need for batching.

So feel free to close the PR.